File: client/src/algorithm.rs
//! State representations of recognized encryption algorithms

use {
    borsh::{BorshDeserialize, BorshSerialize},
    solana_sdk::program_error::ProgramError,
    spl_discriminator::{ArrayDiscriminator, SplDiscriminate},
    spl_keyring_program::tlv::{
        KeystoreEntry, KeystoreEntryConfig, KeystoreEntryConfigEntry, KeystoreEntryKey,
    },
};

/// A trait for defining recognized encryption algorithms
pub trait EncryptionAlgorithm: BorshDeserialize + BorshSerialize + SplDiscriminate {
    /// Returns the key
    fn key(&self) -> Vec<u8>;
    /// Returns the config as an `Option<KeystoreEntryConfig>`
    fn keystore_entry_config(&self) -> Option<KeystoreEntryConfig>;
    /// Converts an encryption algorithm to a buffer
    fn to_buffer(&self) -> Result<Vec<u8>, ProgramError> {
        let mut buffer = Vec::new();
        self.serialize(&mut buffer)?;
        Ok(buffer)
    }
    /// Converts an encryption algorithm to a keystore entry
    fn to_keystore_entry(&self) -> Result<KeystoreEntry, ProgramError> {
        KeystoreEntry::new(
            KeystoreEntryKey {
                discriminator: Self::SPL_DISCRIMINATOR,
                key: self.key(),
            },
            self.keystore_entry_config(),
        )
    }
}

/// A trait representing the configurations of an encryption algorithm
pub trait Configurations: BorshDeserialize + BorshSerialize {
    /// Converts configurations to a buffer
    fn to_buffer(&self) -> Result<Vec<u8>, ProgramError>;
    /// Converts configurations to a `KeystoreEntryConfig`
    fn to_keystore_entry_config(&self) -> Option<KeystoreEntryConfig>;
}

/// Struct representing "no configurations" required for a particular encryption
/// algorithm
#[derive(Clone, Debug, Default, PartialEq, BorshDeserialize, BorshSerialize)]
pub struct NoConfigurations;

impl Configurations for NoConfigurations {
    fn to_buffer(&self) -> Result<Vec<u8>, ProgramError> {
        Ok(vec![0])
    }
    fn to_keystore_entry_config(&self) -> Option<KeystoreEntryConfig> {
        None
    }
}

/// Curve25519 encryption algorithm
#[derive(Clone, Debug, PartialEq, BorshDeserialize, BorshSerialize, SplDiscriminate)]
#[discriminator_hash_input("spl_keyring_program:key:Curve25519")]
pub struct Curve25519([u8; 32]);
impl Curve25519 {
    /// Create a new instance of Curve25519
    pub fn new(key: [u8; 32]) -> Self {
        Self(key)
    }
}

impl EncryptionAlgorithm for Curve25519 {
    fn key(&self) -> Vec<u8> {
        self.0.to_vec()
    }

    fn keystore_entry_config(&self) -> Option<KeystoreEntryConfig> {
        NoConfigurations::default().to_keystore_entry_config()
    }
}

/// Rsa encryption algorithm
#[derive(Clone, Debug, PartialEq, BorshDeserialize, BorshSerialize, SplDiscriminate)]
#[discriminator_hash_input("spl_keyring_program:key:RSA")]
pub struct Rsa([u8; 64]);
impl Rsa {
    /// Create a new instance of Rsa
    pub fn new(key: [u8; 64]) -> Self {
        Self(key)
    }
}

impl EncryptionAlgorithm for Rsa {
    fn key(&self) -> Vec<u8> {
        self.0.to_vec()
    }

    fn keystore_entry_config(&self) -> Option<KeystoreEntryConfig> {
        NoConfigurations::default().to_keystore_entry_config()
    }
}

/// ComplexAlgorithm encryption algorithm
#[derive(Clone, Debug, Default, PartialEq, BorshDeserialize, BorshSerialize, SplDiscriminate)]
#[discriminator_hash_input("spl_keyring_program:key:ComplexAlgorithm")]
pub struct ComplexAlgorithm {
    key: [u8; 32],
    config: ComplexAlgorithmConfigurations,
}
impl ComplexAlgorithm {
    /// Create a new instance of ComplexAlgorithm
    pub fn new(key: [u8; 32], nonce: [u8; 12], aad: [u8; 12]) -> Self {
        Self {
            key,
            config: ComplexAlgorithmConfigurations { nonce, aad },
        }
    }
}

impl EncryptionAlgorithm for ComplexAlgorithm {
    fn key(&self) -> Vec<u8> {
        self.key.to_vec()
    }

    fn keystore_entry_config(&self) -> Option<KeystoreEntryConfig> {
        self.config.to_keystore_entry_config()
    }
}

/// ComplexAlgorithm configurations
#[derive(Clone, Debug, Default, PartialEq, BorshDeserialize, BorshSerialize, SplDiscriminate)]
#[discriminator_hash_input("spl_keyring_program:configuration:ComplexAlgorithm")]
pub struct ComplexAlgorithmConfigurations {
    /// The nonce used for encryption
    pub nonce: [u8; Self::NONCE_LENGTH],
    /// The additional authenticated data
    pub aad: [u8; Self::AAD_LENGTH],
}

impl ComplexAlgorithmConfigurations {
    /// The length of the nonce in bytes
    const NONCE_LENGTH: usize = 12;
    /// The length of the additional authenticated data in bytes
    const AAD_LENGTH: usize = 12;
}

impl Configurations for ComplexAlgorithmConfigurations {
    fn to_buffer(&self) -> Result<Vec<u8>, ProgramError> {
        let mut buffer = Vec::new();
        self.serialize(&mut buffer)?;
        Ok(buffer)
    }

    fn to_keystore_entry_config(&self) -> Option<KeystoreEntryConfig> {
        // 8 Bytes
        let nonce_discriminator = {
            let mut buffer = [0; 8];
            b"nonce".iter().enumerate().for_each(|(i, byte)| {
                buffer[i] = *byte;
            });
            ArrayDiscriminator::new(buffer)
        };
        // 8 Bytes
        let aad_discriminator = {
            let mut buffer = [0; 8];
            b"aad".iter().enumerate().for_each(|(i, byte)| {
                buffer[i] = *byte;
            });
            ArrayDiscriminator::new(buffer)
        };
        Some(KeystoreEntryConfig(vec![
            KeystoreEntryConfigEntry {
                key: nonce_discriminator,
                value: self.nonce.to_vec(),
            },
            KeystoreEntryConfigEntry {
                key: aad_discriminator,
                value: self.aad.to_vec(),
            },
        ]))
    }
}

File: js/src/algorithm.ts
import { Buffer } from "buffer";
import { KeystoreEntry, KeystoreEntryConfig } from "./tlv";

// First 8 bytes of the string literal: "spl_keyring_program:keystore_entry"
export const KEYSTORE_ENTRY_DISCRIMINATOR: Buffer = Buffer.from([
  22, 52, 242, 31, 193, 53, 26, 243,
]);
// First 8 bytes of the string literal: "spl_keyring_program:keystore_entry:configuration"
export const HAS_CONFIGURATIONS_DISCRIMINATOR: Buffer = Buffer.from([
  152, 237, 14, 242, 40, 241, 192, 210,
]);

/**
 * An interface for defining recognized encryption algorithms
 */
export interface EncryptionAlgorithm {
  keyDiscriminator: Buffer;
  keyLength: number;
  key: Buffer;
  config: Configurations;
  toBuffer(): Buffer;
  toKeystoreEntry(): KeystoreEntry;
}

/**
 * Converts an encryption algorithm to a buffer
 * @param algorithm An encryption algorithm
 * @returns A buffer
 */
export function fromEncryptionAlgorithmToBuffer(
  algorithm: EncryptionAlgorithm,
): Buffer {
  return Buffer.concat([
    Buffer.from(algorithm.keyDiscriminator),
    Buffer.from(Uint32Array.of(algorithm.keyLength)),
    Buffer.from(algorithm.key),
    algorithm.config.toBuffer(),
  ]);
}

/**
 * Converts an encryption algorithm to a keystore entry
 * @param algorithm An encryption algorithm
 * @returns A keystore entry
 */
export function fromEncryptionAlgorithmToKeystoreEntry(
  algorithm: EncryptionAlgorithm,
): KeystoreEntry {
  return {
    key: {
      discriminator: algorithm.keyDiscriminator,
      keyLength: algorithm.keyLength,
      key: algorithm.key,
    },
    config: algorithm.config.toKeystoreEntryConfig(),
  };
}

/**
 * Interface representing the configurations of an encryption algorithm
 */
export interface Configurations {
  configurationDiscriminator: Buffer;
  toBuffer(): Buffer;
  toKeystoreEntryConfig(): KeystoreEntryConfig | undefined;
}

/**
 * Class representing "no configurations" required for a
 * particular encryption algorithm
 */
class NoConfigurations implements Configurations {
  configurationDiscriminator: Buffer = NO_CONFIGURATIONS_DISCRIMINATOR;
  constructor() {}
  toBuffer(): Buffer {
    return Buffer.from(this.configurationDiscriminator);
  }
  toKeystoreEntryConfig(): KeystoreEntryConfig | undefined {
    return undefined;
  }
}
// Single zero
export const NO_CONFIGURATIONS_DISCRIMINATOR: Buffer = Buffer.from([0]);

/**
 * Curve25519 encryption algorithm
 */
export class Curve25519 implements EncryptionAlgorithm {
  keyDiscriminator: Buffer = CURVE25519_DISCRIMINATOR;
  keyLength: number = 32;
  key: Buffer;
  config: Configurations = new NoConfigurations();
  constructor(key: Buffer) {
    this.key = key;
  }
  toBuffer(): Buffer {
    return fromEncryptionAlgorithmToBuffer(this);
  }
  toKeystoreEntry(): KeystoreEntry {
    return fromEncryptionAlgorithmToKeystoreEntry(this);
  }
}
// First 8 bytes of the string literal: "spl_keyring_program:key:Curve25519"
export const CURVE25519_DISCRIMINATOR: Buffer = Buffer.from([
  91, 118, 136, 53, 132, 35, 78, 142,
]);

/**
 * RSA encryption algorithm
 */
export class RSA implements EncryptionAlgorithm {
  keyDiscriminator: Buffer = RSA_DISCRIMINATOR;
  keyLength: number = 32;
  key: Buffer;
  config: Configurations = new NoConfigurations();
  constructor(key: Buffer) {
    this.key = key;
  }
  toBuffer(): Buffer {
    return fromEncryptionAlgorithmToBuffer(this);
  }
  toKeystoreEntry(): KeystoreEntry {
    return fromEncryptionAlgorithmToKeystoreEntry(this);
  }
}
// First 8 bytes of the string literal: "spl_keyring_program:key:RSA"
export const RSA_DISCRIMINATOR: Buffer = Buffer.from([
  201, 12, 106, 206, 86, 201, 19, 89,
]);

/**
 * ComplexAlgorithm encryption algorithm
 */
export class ComplexAlgorithm implements EncryptionAlgorithm {
  keyDiscriminator: Buffer = COMPLEX_ALGORITHM_DISCRIMINATOR;
  keyLength: number = 32;
  key: Buffer;
  config: Configurations;
  constructor(key: Buffer, config: ComplexAlgorithmConfigurations) {
    this.key = key;
    this.config = config;
  }
  toBuffer(): Buffer {
    return fromEncryptionAlgorithmToBuffer(this);
  }
  toKeystoreEntry(): KeystoreEntry {
    return fromEncryptionAlgorithmToKeystoreEntry(this);
  }
}
// First 8 bytes of the string literal: "spl_keyring_program:key:ComplexAlgorithm"
export const COMPLEX_ALGORITHM_DISCRIMINATOR: Buffer = Buffer.from([
  238, 108, 0, 133, 126, 20, 221, 160,
]);

/**
 * ComplexAlgorithm configurations
 */
export class ComplexAlgorithmConfigurations implements Configurations {
  configurationDiscriminator: Buffer =
    COMPLEX_ALGORITHM_CONFIGURATION_DISCRIMINATOR;
  // The nonce used for encryption
  nonce: Buffer;
  // The additional authenticated data
  aad: Buffer;
  constructor(nonce: Buffer, aad: Buffer) {
    this.nonce = nonce;
    this.aad = aad;
  }
  toBuffer(): Buffer {
    return Buffer.concat([
      Buffer.from(this.configurationDiscriminator),
      Buffer.from(this.nonce),
      Buffer.from(this.aad),
    ]);
  }
  toKeystoreEntryConfig(): KeystoreEntryConfig {
    return {
      configList: [
        {
          key: Buffer.from("nonce"),
          valueLength: this.nonce.length,
          value: Buffer.from(this.nonce),
        },
        {
          key: Buffer.from("aad"),
          valueLength: this.aad.length,
          value: Buffer.from(this.aad),
        },
      ],
    };
  }
}
// First 8 bytes of the string literal: "spl_keyring_program:configuration:ComplexAlgorithm"
export const COMPLEX_ALGORITHM_CONFIGURATION_DISCRIMINATOR = Buffer.from([
  62, 5, 140, 55, 241, 136, 249, 202,
]);

/**
 * Reads a keystore entry and converts it to one of the above recognized encryption algorithms
 * @param keystoreEntry A keystore entry
 * @returns An encryption algorithm
 */
export function fromKeystoreEntrytoEncryptionAlgorithm(
  keystoreEntry: KeystoreEntry,
): EncryptionAlgorithm {
  const keyDiscriminator = keystoreEntry.key.discriminator;
  const key = keystoreEntry.key.key;
  if (keyDiscriminator.equals(CURVE25519_DISCRIMINATOR)) {
    return new Curve25519(key);
  } else if (keyDiscriminator.equals(RSA_DISCRIMINATOR)) {
    return new RSA(key);
  } else if (keyDiscriminator.equals(COMPLEX_ALGORITHM_DISCRIMINATOR)) {
    const config = keystoreEntry.config;
    if (!config) {
      throw new Error("Missing required config for ComplexAlgorithm");
    }
    const configList = config.configList;
    const nonce = configList[0].value;
    const aad = configList[1].value;
    return new ComplexAlgorithm(
      key,
      new ComplexAlgorithmConfigurations(nonce, aad),
    );
  } else {
    throw new Error("Unrecognized key discriminator");
  }
}
